OpenSim/Common/Component.h

```diff
@@ -4294,8 +4294,9 @@ void ComponentListIterator<T>::advanceToNextValidComponent() {
     // Advance _node to next valid (of type T) if needed
     // Similar logic to operator++ but applies _filter->isMatch()
     while (_node != nullptr && (dynamic_cast<const T*>(_node) == nullptr ||
-                                !_filter.isMatch(*_node) ||
-                                (_node == _root))){
+                                !(!_filter || _filter->isMatch(*_node)) ||
```

Comment:
This change is to accomodate the fact that `_filter` is now optional (nullable) - it's an extra branch vs the original code but prevents the implementation from hopping into a virtual function, heap-allocating a clone of a filter object, etc.